### PR TITLE
[inductor] fix incorrect strides in copy() decomp, fix hf_LongFormer + hf_BigBird errors

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/100067, https://github.com/pytorch/pytorch/issues/98268 and https://github.com/pytorch/pytorch/issues/93428.

See the comment [here](https://github.com/pytorch/pytorch/issues/100067#issuecomment-1523856970) for details. The bug was that the decomposition that inductor uses for `aten.copy` doesn't respect the strides of the input in all cases. The fixes that I added should work, but will be pretty slow - we allocate a tensor (potentially larger than `self` if `self` is a slice), and perform an `as_strided_scatter` + `as_strided`. Longer term, stride-agnostic IR should let us remove this decomp?  cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @anijain2305 @soumith @desertfire 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104476
* #104793
* __->__ #100115

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/decomposition.py
Comment (albanD) [inline torch/_inductor/decomposition.py:362]: You shoulds check storage offset as well here I think?
Also in all the conditions below as contiguous =/> storage_offset == 0

File: torch/_inductor/decomposition.py
Comment (albanD) [inline torch/_inductor/decomposition.py:366]: It is ok to return `src` as-is here?

File: torch/_inductor/decomposition.py
Comment (albanD) [inline torch/_inductor/decomposition.py:372]: Add assert that is_neg/is_conj is false?

File: torch/_inductor/decomposition.py
Comment (albanD) [inline torch/_inductor/decomposition.py:374]: Why is this not `self.untyped_storage().size()` ?

File: torch/_inductor/decomposition.py
Comment (albanD) [inline torch/_inductor/decomposition.py:378]: You need to fill in the buffer with the content of self at some point here?

**Comment by ngimel** [comment]
Hm, I thought for graph outputs we are setting strides to fx node meta, so even if copy itself returns something wrong, the graph output should be made right? cc @eellison

**Comment by eellison** [comment]
@ngimel we do fix the output strides but we use fake tensor to get them. if the decomp for copy gives wrong strides we'll be wrong too

File: torch/_inductor/decomposition.py
Comment (bdhirsh) [reply torch/_inductor/decomposition.py:365]: I added this case for if `self` is contiguous, but `src` isn't. So I don't think returning `src` as-is here is ok.

File: torch/_inductor/decomposition.py
Comment (bdhirsh) [reply torch/_inductor/decomposition.py:375]: hmm, so... do I?

I want to make the argument that this memory should never get read (and the **only** reason we need to allocate it is to keep strides correct so our ops that aren't stride-agnostic are happy, like view).

The example where I think you could imagine that data being accessed is in code like this:
```
a = torch.zeros(2, 2)
a_slice = a[:, 1]
a.copy_(toch.ones_like(a_slice))
# What happens if the user accesses other parts of `a`, that are not `a_slice`?
```

But that case should be fine - any accesses of `a` (or any other views of `a` should be turned into fresh tensors via `slice/select_scatter` call, which will ignore the unitialized data when it's scattering the updated data into the larger tensor.

The other case I can think of is using `as_strided()` to directly access memory outside the bounds of the current slice, but this is banned in functionalization anyway.

File: torch/_inductor/decomposition.py
Comment (albanD) [reply torch/_inductor/decomposition.py:365]: Sorry I meant in the sense of "native functions should never return their input as-is and at least do a self.alias()".

**Review by albanD — APPROVED** [review]
Clarified offline with Brian. All sounds good!

**Comment by ngimel** [comment]
Would it make sense to run perf test on it before landing?

**Comment by bdhirsh** [comment]
Added `ciflow/inductor-perf-compare` (I hope that's the right label)

**Comment by bdhirsh** [comment]
I'm also getting a few cursed errors in CI, need to investigate. e.g.:
```
RuntimeError: 32*s0*s1*s2 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7fccc1335a50>
```

**Comment by ngimel** [comment]
Should we try the lowering approach then, in the hopes it'd help the tests?

**Comment by bdhirsh** [comment]
I ran into some more issues trying to beef up the decomp: in particular, it looks like all of the different subclasses of `BaseView`/`IRNode` would need to have `get_stride()` / `get_storage_offset()` methods implemented on them (this seems non-trivial?)


> Should we try the lowering approach then, in the hopes it'd help the tests?

Sounds good, I'm starting to try that now. The immediate problem is that `.to()` (and also `.expand_copy()`) don't have lowerings, and we can no longer use their decompositions - because the decompositions expect to be able to take in tensors, but we're now passing in IRNodes to these ops.

I'm going to try to see if I can emulate the behavior of `.to()` using inductor lowering primitives - does that sound like a feasible goal?

File: torch/_inductor/lowering.py
Comment (bdhirsh) [inline torch/_inductor/lowering.py:1582]: cc @ngimel does this lowering for `copy()` look reasonable to you?

Comment (bdhirsh) [reply torch/_inductor/lowering.py:1582]: There's one test that's not passing: `test/inductor/test_torchinductor_dynamic_shapes.py DynamicShapesCpuTests.test_input_mutation2_dynamic_shapes_cpu`

```
        def fn(a):
            b = a + 1
            a.view(64).copy_(torch.tensor([66.0], device=a.device))
            c = a + 2
            return b, c
```

It fails with this:
```
AttributeError: 'Pointwise' object has no attribute 'get_stride'
```

And if I print the graph that inductor is trying to lower, I see:
```
def forward(self, primals_1):
    clone = torch.ops.aten.clone.default(primals_1)
    add = torch.ops.aten.add.Tensor(clone, 1)
    view = torch.ops.aten.view.default(clone, [64])
    _tensor_constant0 = self._tensor_constant0
    lift_fresh_copy = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0);  _tensor_constant0 = None
    copy = torch.ops.aten.copy.default(view, lift_fresh_copy);  view = lift_fresh_copy = None
    sym_size = torch.ops.aten.sym_size(primals_1, 1);  primals_1 = None
    view_1 = torch.ops.aten.view.default(copy, [1, sym_size]);  copy = None
    add_1 = torch.ops.aten.add.Tensor(view_1, 2)
    sym_stride = torch.ops.aten.sym_stride(clone, 0);  clone = None
    return [view_1, add, add_1, sym_size, sym_stride]
```

Should we not expect to see a `sym_stride()` call in the inductor graph? I can dig through the test some more to see where it's coming from.

Comment (ezyang) [reply torch/_inductor/lowering.py:1582]: I think this is some preexisting problem; in fact, hf_LongFormer is failing on dynamic shapes on master for exactly the same reason

Comment (ngimel) [reply torch/_inductor/lowering.py:1582]: Hm I agree it's a preexisting problem, I haven't seen `sym_stride` in the inductor graph previously, perhaps we should register lowering for it.

Comment (ngimel) [reply torch/_inductor/lowering.py:1582]: Upd: we already have a lowering for `sym_stride`, the problem is that `clone` is not a realized buffer, and thus `strides` for it are not defined. Should we instead return strides from the fx graph meta data?

Comment (ngimel) [reply torch/_inductor/lowering.py:1582]: Getting the result from meta makes the test pass, add 
```
            elif n.target == torch.ops.aten.sym_stride:
                if isinstance(n.meta["val"], torch.SymInt):
                    result = n.meta["val"].node.expr
                else:
                    result = super().run_node(n)
```
in `run_node` in graph.py

Comment (ezyang) [reply torch/_inductor/lowering.py:1582]: @bdhirsh Can you post the full backtrace here?

Comment (bdhirsh) [reply torch/_inductor/lowering.py:1582]: Hmm, looks like we are saving the sym_stride so that we can use it in the backward (via an `empty_strided()` call):
```
def forward(self, sym_size, sym_stride, tangents_1, tangents_2, tangents_3):
    add_2 = torch.ops.aten.add.Tensor(tangents_1, tangents_3);  tangents_1 = tangents_3 = None
    new_empty_strided = torch.ops.aten.new_empty_strided.default(add_2, [1, sym_size], [sym_stride, 1]);  sym_stride = None
    copy_1 = torch.ops.aten.copy.default(new_empty_strided, add_2);  new_empty_strided = add_2 = None
    view_5 = torch.ops.aten.view.default(copy_1, [64]);  copy_1 = None
    clone_1 = torch.ops.aten.clone.default(view_5, memory_format = torch.contiguous_format)
    full_like = torch.ops.aten.full_like.default(clone_1, 0, pin_memory = False, memory_format = torch.contiguous_format);  clone_1 = None
    copy_2 = torch.ops.aten.copy.default(view_5, full_like);  view_5 = full_like = None
    view_6 = torch.ops.aten.view.default(copy_2, [1, sym_size]);  copy_2 = sym_size = None
    add_3 = torch.ops.aten.add.Tensor(view_6, tangents_2);  view_6 = tangents_2 = None
    return [add_3]
```

It looks like that's coming from autograd's `CopySlices`: https://github.com/pytorch/pytorch/blob/01476465ddb17824d00c9f5dbc39daf984f2267c/torch/csrc/autograd/functions/tensor.cpp#L79

Comment (bdhirsh) [reply torch/_inductor/lowering.py:1582]: It seems like we're stuck with the possibility of saving strides for the backwards then? Unless we want to special case autograd's CopyBackwards decomposition for inductor.

Although even if we end up passing incorrect strides into the backward graph, does that really matter in practice? (since inductor is stride agnostic). That would mean the only remaining problem is "can we return output strides from an inductor forward graph without inductor complaining".

Comment (ezyang) [reply torch/_inductor/lowering.py:1582]: It should be fine for us to propagate strides from forward to backward. Inductor can take the strides and ignore it (because it doesn't need to know about strides, only what the final striding is).

File: torch/_inductor/graph.py
Comment (ezyang) [inline torch/_inductor/graph.py:497]: occasionally

**Comment by ngimel** [comment]
AllenaiLongformerBase likely is broken because of this bug, hopefully this will fix it [hud.pytorch.org/benchmark/huggingface/inductor_no_cudagraphs?startTime=Wed%2C%2003%20May%202023%2023%3A29%3A52%20GMT&stopTime=Wed%2C%2010%20May%202023%2023%3A29%3A52%20GMT&mode=training&dtype=amp&lBranch=main&lCommit=d719f0276d69a8315b65f4c4500cfc1cdaddb025&rBranch=main&rCommit=675029aabf143a60aee62c21e093ed1091f46743&model=AllenaiLongformerBase](http://hud.pytorch.org/benchmark/huggingface/inductor_no_cudagraphs?startTime=Wed%2C%2003%20May%202023%2023%3A29%3A52%20GMT&stopTime=Wed%2C%2010%20May%202023%2023%3A29%3A52%20GMT&mode=training&dtype=amp&lBranch=main&lCommit=d719f0276d69a8315b65f4c4500cfc1cdaddb025&rBranch=main&rCommit=675029aabf143a60aee62c21e093ed1091f46743&model=AllenaiLongformerBase)

**Comment by bdhirsh** [comment]
I looked through the CI failures, and they looked flaky (`resnet50_quantized_qat` was failing, but I could repro on a different commit before this PR). Re-running CI.

**Comment by bdhirsh** [comment]
Internal failure was because the inductor config `disable_cpp_codegen` is only set to True in internal, and that wasn't playing well with `lift_fresh_copy()` (thanks @ngimel for the help)! I followed Natalia's suggestion of ensuring that we never try to "fall back" on `lift_fresh_copy()` even when `disable_cpp_codegen` is set, which fixes that test.

**Review by ngimel — APPROVED** [review]

**Comment by bdhirsh** [comment]
Looks like cpp_codegen tests are failing because with my change, we still end up generating some cpp wrapper code (even if we only "ignore" `lift_fresh_copy`). Example from the failing test:
```
> /scratch/hirsheybar2/work/pytorch/test/inductor/test_cpu_repro.py(901)test_skip_cpp_codegen()
-> FileCheck().check_not("void kernel").run(code)
(Pdb) print(code)
Output code written to: /tmp/torchinductor_hirsheybar/gh/cgh7g2hgldpacxqz54c3euxdklup52ws7gnon47nmwzlvsuim7gd.py
Output code:

from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile

from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()
constant0 = None  # 9802b3c1315eab52d2c262cf819e11e7f511726733be6fa5905701ce5beea92b


cpp_fused_lift_fresh_0 = async_compile.cpp('''
#include "/tmp/torchinductor_hirsheybar/5d/c5dsrjrcd4jlzryilhxl5hdvcrzsoek52xzzqqy57hcoezvxxxwm.h"
extern "C" void kernel(const long* in_ptr0,
                       long* out_ptr0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
}
''')


async_compile.wait(globals())
del async_compile
```

Still looking, but my guess is that I need to **only** do the skip if the value that we're lifting is a scalar (0-dim tensor), and then arrange to make sure we don't generate a cpp kernel for that case.

**Comment by bdhirsh** [comment]
Quick update on this PR:

(1) Making `copy()` a lowering doesn't play well with an existing inductor config, `disable_cpp_codegen=True`

(2) initially, I tried forcing inductor codegen to work around it: the problem was that `lift_fresh_copy()` is technically a "CPU" op, so disable_cpp_codegen says that we should fall back to ATen instead of running it.  If the input happens to be a 0-dim tensor, and we later want to fuse it into cuda tensor compute (in a triton kernel), we **shouldn't** fall back to ATen - the triton codegen is supposed to inline the 0-dim value as a scalar directly into the triton kernel. But if we later want to use the 0-dim tensor with other CPU tensors, then those operators should fall back to ATen (since no cpu compute is allowed). It seems like this isn't very easy to do.

There are some specific things that we could do to get the failing test to pass (that Natalia helpfully pointed out), but they're all pretty fragile and could cause other issues later.

(3) The "real" fix is to kill the `disable_cpp_codegen=True` flag, which is only used internally today.

So this PR is currently blocked on 3 (fixing internal). I'm going to push this off for the rest of the week to focus on landing some other PR's (I'm also on PTO this Thursday and Friday), but I'm thinking of trying to kill that config next week if I have time.

**Comment by bdhirsh** [comment]
diff to kill `inductor.config.disable_cpp_codegen` landed (https://github.com/pytorch/pytorch/pull/104351), so I can finally land this (cc @anijain2305)

**Comment by bdhirsh** [comment]
This PR is unblocked now that `inductor.config.disable_cpp_codegen` is turned off internally (there was a hotfix that I had to land end of last week).

But I see a CI failure on this PR that looks real, coming from `yolov3`. Need to dig into it more:
```
2023-07-07T20:59:24.0781472Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 1895, in slice_scatter
2023-07-07T20:59:24.0781786Z     assert x.get_dtype() == src.get_dtype()
2023-07-07T20:59:24.0782137Z torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2023-07-07T20:59:24.0782421Z LoweringException: AssertionError: 
2023-07-07T20:59:24.0782666Z   target: aten.slice_scatter.default
```

**Comment by bdhirsh** [comment]
torchbench failures have been fixed. There's still one failing test, which is a segfault in `test/test_torch.py`. I haven't been able to repro this locally so far.

**Comment by bdhirsh** [comment]
I was able to repro the segfault locally - it looks like by turning the decomp into a lowering, we are now relying on the meta function for `aten.copy()` to do proper error checking (which it doesn't). The test passed an invalidly strided input into `copy()`, which inductor faithfully tries to codegenerate a kernel for that segfaults.

File: torch/_meta_registrations.py
Comment (bdhirsh) [inline torch/_meta_registrations.py:290]: @ezyang does this meta function look ok to you? Replicating all the error checking in our `copy_()` kernel is pretty annoying - I figure it's pretty harmless to do this since we're running with meta tensors.

Comment (bdhirsh) [reply torch/_meta_registrations.py:290]: nvm this doesn't work (when I wrote it I thought I was writing the meta impl for `aten.copy`, not `aten.copy_`. But this will just infinite loop).

Hmm, maybe we'll need to replicate some of the meta checks in the C++ `copy_()` impl (or figure out a way to call it here - since it's `CompositeExplicitAutograd`)

**Pull request closed without merging**
